Delegatecall/Callcode CREATE from account with max nonce tests

marioevz opened the pull request:
This PR implements test cases where an account with the max nonce (0xffffffffffffffff) delegate calls to another contract that attempts to create a contract, which should fail due to the max nonce being reached.

winsvega commented:
using the test cases from here  (first 8 cases)
https://github.com/ethereum/tests/issues/991

can you make more comments around which case is covered from the list.
you can also use `:label mark`  in data for the expect sections instead of indexes. 
and the expect section better be separate when possible.

winsvega commented on src/GeneralStateTestsFiller/stCreateTest/CREATE_HighNonceMinus1DelegatecallFiller.yml:
```diff
@@ -0,0 +1,75 @@
+      network:
+        - '>=Homestead'
+      result:
+       a94f5374fce5edbc8e2a8697c15331677e6ebf0b:
```
sice this also covers the other case, here must be expect: shouldnotxist the account that would have been created if the nonce rule worked. (you can calculate it for create2) 
check the nonce value of accounts b94, c0dec
I think the expect section better be separate

marioevz replied:
Added appropriate 'shouldnotexist' tests in latest push.

winsvega commented on src/GeneralStateTestsFiller/stCreateTest/CREATE_HighNonceMinus1DelegatecallFiller.yml:
```diff
@@ -0,0 +1,75 @@
+
+  transaction:
+    data:
+      - '0x00'
```
you can use `:label mark` to reference the data in expect sections

marioevz replied:
Used proper labels for all test cases in latest push.

marioevz commented:
- Updated test to much better description.
- Added tests for CREATE2
- Single test file for MaxNonce and MaxNonce - 1
- Test cases covered from #991 are 2 through 9.

marioevz commented on src/GeneralStateTestsFiller/stCreateTest/CREATE_HighNonceDelegatecallFiller.yml:
```diff
@@ -0,0 +1,177 @@
+              pop(call(sub(gas(), 1000), callcontract, 0, 0, 0, 0, 32))
+              let result := mload(0)
+              sstore(i, result)
+              pop(delegatecall(sub(gas(), 1000), callcontract, 0, 0, 0, 32))
```
This delegate call allows us to increase Contract A nonce to MaxNonce and test:
A (maxNonce) calls B (maxNonce), contract creation fails.
Same for CREATE2.

winsvega replied:
I think here it gets too complicated. when a contract do 3 times (call->create then delegatecall->create).

marioevz replied:
I removed this complicated scenario by separating into two files for CREATE, and two files for CREATE2.
First file contains origin contract nonce = 0xfffffffffffffffe, second one nonce = 0xffffffffffffffff.

winsvega commented on src/GeneralStateTestsFiller/stCreateTest/CREATE_HighNonceDelegatecallFiller.yml:
```diff
@@ -0,0 +1,187 @@
+    # - Contract A (Nonce=0xfffffffffffffffe) delegate calls Contract B (Nonce=0xffffffffffffffff), contract created first call, then fails
+      - ':label a_max_nonce_delegate_b_max_nonce :abi f(uint,uint,uint) 0x00 0xffffffffffffffff 0x03'
+    # - Contract A (Nonce=0xfffffffffffffffe) code calls Contract B (Nonce=0xfffffffffffffffe), contract created first call, then fails
+      - ':label a_max_nonce_delegate_b_max_nonce_minus_1 :abi f(uint,uint,uint) 0x01 0xfffffffffffffffe 0x03'
```
I thought 0x01 - means callcode ?  the label remains delegate

marioevz replied:
Changed comments in latest push, let me know if it's ok.

winsvega commented on src/GeneralStateTestsFiller/stCreateTest/CREATE_HighNonceDelegatecallFiller.yml:
```diff
@@ -0,0 +1,187 @@
+    # Parameters:
+    #  1) Call type: 0 - delegateCall, 1 - callCode, 2 - call + delegate call, 3 - delegate call + call
+    #  2) Called contract nonce
+    #  3) Contracts to create
```
this parameter is never used. it stays the same 0x03

marioevz replied:
Removed in latest push.

winsvega commented on src/GeneralStateTestsFiller/stCreateTest/CREATE_HighNonceDelegatecallFiller.yml:
```diff
@@ -0,0 +1,187 @@
+    #  1) Call type: 0 - delegateCall, 1 - callCode, 2 - call + delegate call, 3 - delegate call + call
+    #  2) Called contract nonce
+    #  3) Contracts to create
+    # - Contract A (Nonce=0xfffffffffffffffe) delegate calls Contract B (Nonce=0xfffffffffffffffe), contract created first call, then fails
```
then why it is `a_max_nonce_delegate_b_max_nonce_minus_1`
if nonce is the same

marioevz replied:
Changed labels in latest push.

winsvega commented on src/GeneralStateTestsFiller/stCreateTest/CREATE_HighNonceDelegatecallFiller.yml:
```diff
@@ -0,0 +1,187 @@
+  transaction:
+    data:
+    # Parameters:
+    #  1) Call type: 0 - delegateCall, 1 - callCode, 2 - call + delegate call, 3 - delegate call + call
```
why you need case 2 and 3 ?

marioevz replied:
Removed these unnecessary test cases and split into two files to make it more readable and have all the coverage.

winsvega commented on src/GeneralStateTestsFiller/stCreateTest/CREATE_HighNonceDelegatecallMinus1Filler.yml:
```diff
@@ -0,0 +1,156 @@
+# The test calls CREATE from an account with max allowed nonce - 1.
```
this you can put into
```
 _info:     
    comment: ""
```

sectino this way it will pring in generated test for the devs to read

winsvega commented on src/GeneralStateTestsFiller/stCreateTest/CREATE_HighNonceDelegatecallMinus1Filler.yml:
```diff
@@ -0,0 +1,156 @@
+    #       2 - call
+    #  2) Called contract nonce
+    # - Contract A (Nonce=0xfffffffffffffffe) delegate calls Contract B (Nonce=0xfffffffffffffffe), contract create first suceeds, then fails
+      - ':label a_max_nonce_minus_1_delegate_call_create :abi f(uint,uint) 0x00 0xfffffffffffffffe'
```
you can make it even more straight foward. 

```
1)  call type
2)  caller nonce  (A)
3)  destination nonce (B)
4) create/create2 for the destiantion
```

reveiver contract parse the args.  (transaction destiantion)
and make a call to callernonce (A) contact. passing the calltype and destination and create type as argument
then callernonce contract (A) do the calltype to destination (B)  and destination (B) do the create/create2
after each operation to mark success callernonce(A)  stores 1:1,  destination B stores 2:2 to check in expect section
also check nonce in expect section for both all variants of A, and B, and storage at cells 1,2 for all types of A and B

marioevz replied:
I did all these changes in latest push with some minor differences:
- callernonce (A) is always the same contract, but for the test cases where callernonce == max nonce, the contract first does a dummy CREATE to increase its nonce before doing any delegate call/call code/call.
- The store actions store the resulting contract address instead of a constant 1 or 2.

Let me know if this is ok.

winsvega commented on src/GeneralStateTestsFiller/stCreate2/CREATE2_HighNonceDelegatecallFiller.yml:
```diff
@@ -0,0 +1,588 @@
+          let createtype    := calldataload(0)
+          let contextnonce  := sload(0xffff)
+
+          // initcode: { return(0, 1) }
```
Can also track the existance of created contracts.
If init code puts smth to storage you will see the address in post state and would be able to check the storage

marioevz replied:
Added sstore(1, 1) to created contracts' code, call all contracts after creation, and verify the storage in each contract.

winsvega commented on src/GeneralStateTestsFiller/stCreate2/CREATE2_HighNonceDelegatecallFiller.yml:
```diff
@@ -0,0 +1,588 @@
+            // We have a lower nonce than required for the caller, create dummy contract to increase nonce
+            mstore(0, 0x60016000f3)
+            let addr := create(0, sub(32, 5), 5)
+            if gt(addr, 0) { sstore(0xffff, add(contextnonce, 1)) }
```
Else?

marioevz replied:
I think else might be unnecessary here, since the store value is supposed to mimic the nonce and the nonce is not increased if the address returned is zero (contract creation failed).

winsvega commented on src/GeneralStateTestsFiller/stCreate2/CREATE2_HighNonceDelegatecallFiller.yml:
```diff
@@ -0,0 +1,588 @@
+  expect:
+    - indexes:
+       data:
+       - ':label a_max_nonce_delegate_call_create'
```
Perhaps better `A_MaxNonce_DelegateCall_B_MaxNonce_Create`

marioevz replied:
Change the labels for all test cases to this format.

winsvega commented on src/GeneralStateTestsFiller/stCreate2/CREATE2_HighNonceDelegatecallFiller.yml:
```diff
@@ -0,0 +1,588 @@
+    - indexes:
+       data:
+       - ':label a_max_nonce_delegate_call_create'
+       - ':label a_max_nonce_call_code_create'
```
It is unclear to which nonce contract A is making a call

marioevz replied:
Added a better labeling format.

marioevz replied:
Added a better labeling format.

winsvega commented on src/GeneralStateTestsFiller/stCreate2/CREATE2_HighNonceDelegatecallFiller.yml:
```diff
@@ -0,0 +1,588 @@
+          nonce: '0xfffffffffffffffe'
+          storage:
+            '0xffff': '0xfffffffffffffffe'
+       000000000000000000000000ffffffffffffffff:
```
Put a comment that this is a destination B with max nonce

marioevz replied:
Placed comments in all verified addresses that describe each contract.

winsvega commented:
ok, I think now it looks much better
